docs/software_requirements/semaphore.sdoc

```diff
@@ -0,0 +1,354 @@
+TITLE: Dynamic Initialization
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism to define and initialize a semaphore at runtime.
+
+----
+
+The Zephyr RTOS shall expose an interface to dynamically initialize a Counting Semaphore which requires the following information:
+
+- initial count of the Counting Semaphore;
+- count limit of the Counting Semaphore.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-SEMAPHORE-003
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Maximum limit of a semaphore
+STATEMENT: >>>
+When a semaphore is used for counting purposes and when the semaphore does not have an explicit
+maximum limit, the Zephyr RTOS shall define the maximum limit of a semaphore to
+be used when the semaphore is initialized.
+<<<
+
+[/SECTION]
+
+[SECTION]
+TITLE: Take semaphore
+
+[REQUIREMENT]
+UID: ZEP-SEMAPHORE-005
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore acquisition
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism allowing threads to take (acquire) a semaphore, ensuring atomic and thread-safe operations for
+semaphore acquisition.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-SEMAPHORE-007
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore Timeout
+STATEMENT: >>>
+When attempting to take (acquire) a semaphore, the Zephyr RTOS shall accept options that specify timeout periods, allowing threads to set a maximum wait time for semaphore acquisition.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[/SECTION]
+
+[SECTION]
+TITLE: Give semaphore
+
+[REQUIREMENT]
+UID: ZEP-SEMAPHORE-010
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore release
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism allowing threads to give (release) a semaphore, ensuring atomic and thread-safe operations for
+semaphore release.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[/SECTION]
+
+[/SECTION]
+
+[SECTION]
+TITLE: Functional requirements
+
+[SECTION]
+TITLE: Initialize semaphore
+
+[REQUIREMENT]
+UID: ZEP-SEMAPHORE-004
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Initialialization with inital semaphore value
+STATEMENT: >>>
+When initializing a counting semaphore, the initial semaphore value shall be set.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-SEMAPHORE-016
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Initialialization with maximum count value
+STATEMENT: >>>
+When initializing a counting semaphore, the maximum permitted count a semaphore
+can have shall be set.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-SEMAPHORE-017
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore Initialization Option Validation
+STATEMENT: >>>
+When initializing a counting semaphore, where the maximum permitted count of a semaphore is invalid,
+the Zephyr RTOS shall return an error indicating invalid values.
```

Comment:
**nitpick:** return an 'invalid value' error. We probably need to add to the glossary the possible error values at some later time.

Reply:
FYI, [this PR](https://github.com/zephyrproject-rtos/reqmgmt/pull/60) provides a glossary document skeleton and some definitions from standards.